package/kernel/mac80211/patches/ath11k/453-ath11k-add-ath11k_mac_op_flush_sta-to-properly-flush.patch

```diff
@@ -0,0 +1,49 @@
+From 5cff5dfbb350af519d8be49ca4cc6a20c54d8afa Mon Sep 17 00:00:00 2001
+From: Florian Maurer <f.maurer@outlook.de>
+Date: Fri, 3 Oct 2025 12:56:13 +0200
+Subject: [PATCH] ath11k: add ath11k_mac_op_flush_sta to properly flush pending
+ packets
+
+warning print "ath11k c000000.wifi: failed to flush transmit queue 0"
+is observed on various openwrt devices.
+
+Signed-off-by: Florian Maurer <f.maurer@outlook.de>
+---
+ drivers/net/wireless/ath/ath11k/mac.c | 19 +++++++++++++++++++
+ 1 file changed, 19 insertions(+)
+
+--- a/drivers/net/wireless/ath/ath11k/mac.c
++++ b/drivers/net/wireless/ath/ath11k/mac.c
+@@ -8248,6 +8248,24 @@ static void ath11k_mac_op_flush(struct i
+ 	ath11k_mac_flush_tx_complete(ar);
+ }
+ 
++static void ath11k_mac_op_flush_sta(struct ieee80211_hw *hw,
++							struct ieee80211_vif *vif,
++			    			struct ieee80211_sta *sta)
++{
++	struct ath11k_vif *arvif = (void *)vif->drv_priv;
++	struct ath11k *ar = hw->priv;
++	struct peer_flush_params params = {
++		.peer_tid_bitmap = 0xFF,
```

Comment:
This works fine with `0xFF` and with `0xFFFF`, but not with `0xFFFFFFFF` - so there probably is an undocumented mgmt bit which should not be set as it would issue `Assertion !CHK_TID_FLG(ptid, WAL_TID_IN_SCHEDQ) failedparam0 :zero, param1 :zero, param2 :zero.`
I don't quite know what the other bits do here.

For ath10k, the mgmt bit is `WMI_MGMT_TID=17`, which is not the case here.

Reply:
Management frame goes through a different path

Reply:
We do not really know what each bit means. So I would say that there is the theoretical chance that we might not actually be flushing everything. The obvious thing would be that the TIDs are mapped into the bottom bits, and then things should be fine as we only care about data frames.

Note that we don't even need the flush to complete from the driver perspective. All we really care about is that the FW flushes before processing the next command (which is the key deletion).

It would be good to have some kind of confirmation from someone who can inspect the FW. It might be that we technically need to wait for the flush to complete by waiting for the TX reports in the driver (same on ath10k).